File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, a: List[int], t: int) -> List[List[int]]:
        a.sort()
        b,c=[],[]
        self.brr(a,0,b,t,c)
        return c
    
    def brr(self,a,i,b,k,c):
        if i==len(a):
            if k==0:
                c.append(b.copy())
            return
        if a[i]<=k:
            b.append(a[i])
            self.brr(a,i+1,b,k-a[i],c)
            b.pop()
        while i+1<len(a) and a[i]==a[i+1]:
            i+=1
        self.brr(a,i+1,b,k,c)

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& a, int t) {
        sort(a.begin(), a.end());
        vector<vector<int>> c;
        vector<int> b;
        brr(a, 0, b, t, c);
        return c;
    }
private:
    void brr(vector<int>& a, int i, vector<int>& b, int k, vector<vector<int>>& c) {
        if (i == a.size()) {
            if (k == 0) c.push_back(b);
            return;
        }
        if (a[i] <= k) {
            b.push_back(a[i]);
            brr(a, i + 1, b, k - a[i], c);
            b.pop_back();
        }
        while (i + 1 < a.size() && a[i] == a[i + 1]) i++;
        brr(a, i + 1, b, k, c);
    }
};